### PR TITLE
feat(rbac): treat the empty ownership as owned by ACCOUNT_ADMIN instead of PUBLIC

### LINKED ISSUE #13032: Feature: limit the system database are owned by account_admin by default
**Summary**

currently the system.tables are available to all the users.

however these tables are better to considered as admin-only.

we can leverage the DAC mechanism to limit the access of the system tables to be `account_admin` only (owned by `account_admin`).

### PR DESCRIPTION
**Pull request by flaneur2020**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

after changing uninitialized ownership as ACCOUNT_ADMIN, the default database would not accessible by users with PUBLIC role without explictly privilege granted.

another point is the rewritten statements like `SHOW TABLES` and `SHOW DATABASES` should still be allowed. so I added an allowlist on the tables under `system` database which is required on these `SHOW xxx` statements to bypass the privilege checks.

this pr also made a small refactor about removing the `verify_ownership` parameter in `validate_table_access` and `validate_db_access`, this parameter is considered to be always true in these cases.

fixes #14163
fixes #13032

## Tests

- [x] Logic Test

## Type of change

- [x] New Feature (non-breaking change which adds functionality)

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/14112)

### CODE REVIEW

File: src/query/service/src/sessions/session_privilege_mgr.rs [hunk 1]
```diff
@@ -269,14 +270,15 @@ impl SessionPrivilegeManager for SessionPrivilegeManagerImpl {
...
+        let owner_role_name = match role_mgr.find_object_owner(&tenant, object).await? {
+            Some(owner_role) => owner_role.name,
+            None => BUILTIN_ROLE_ACCOUNT_ADMIN.to_string(),
         };
 
         let available_roles = self.get_all_available_roles().await?;
-        if !available_roles.iter().any(|r| r.name == owner_role.name) {
+        if !available_roles.iter().any(|r| r.name == owner_role_name) {
             return Err(ErrorCode::PermissionDenied(
                 "Permission denied, current session do not have the ownership of this object"
```
Comment (TCeason) [inline src/query/service/src/sessions/session_privilege_mgr.rs:283]: Add owner_role info into this err msg.

File: tests/suites/0_stateless/18_rbac/20_0012_privilege_access.result [hunk 2]
```diff
@@ -18,7 +18,8 @@ test -- select
 1
 1
 1
-1
+test -- select view
+Error: APIError: ResponseError with 1063: Permission denied, privilege [Select] is required on 'default'.'default2'.'v_t20_0012' for user 'test-user'@'%' with roles [public,test-role1,test-role2]
```
Comment (TCeason) [inline tests/suites/0_stateless/18_rbac/20_0012_privilege_access.result:22]: This modify generates a break change.

Before this pr, user can query view even though he does not have privilige about any table.

File: tests/suites/0_stateless/18_rbac/20_0015_set_role.sh [hunk 3]
```diff
@@ -21,8 +25,8 @@ echo 'GRANT ROLE testrole2 to ROLE testrole3' | $BENDSQL_CLIENT_CONNECT
 echo 'GRANT ROLE testrole1 to testuser1' | $BENDSQL_CLIENT_CONNECT
 echo 'GRANT ROLE testrole2 to testuser1' | $BENDSQL_CLIENT_CONNECT
 echo 'GRANT ROLE testrole3 to testuser1' | $BENDSQL_CLIENT_CONNECT
-echo "CREATE TABLE t20_0015_table1(c int not null)" | $BENDSQL_CLIENT_CONNECT
-echo "CREATE TABLE t20_0015_table2(c int not null)" | $BENDSQL_CLIENT_CONNECT
+echo "CREATE TABLE t20_0015_table1(c int not null) ENGINE = MEMORY" | $BENDSQL_CLIENT_CONNECT
+echo "CREATE TABLE t20_0015_table2(c int not null) ENGINE = MEMORY" | $BENDSQL_CLIENT_CONNECT
```
Comment (TCeason) [inline tests/suites/0_stateless/18_rbac/20_0015_set_role.sh:29]: Need tests/suites/0_stateless/18_rbac/20_0015_set_role.result?

File: tests/suites/0_stateless/18_rbac/20_0016_rewrite_statements.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+echo "DROP USER IF EXISTS 'testuser1'" | $BENDSQL_CLIENT_CONNECT
+
+echo '-- prepare user and tables for tests'
+echo "CREATE USER 'testuser1' IDENTIFIED BY '$TEST_USER_PASSWORD'" | $BENDSQL_CLIENT_CONNECT
+echo "CREATE TABLE IF NOT EXISTS t20_0016_table1(c int not null)" | $BENDSQL_CLIENT_CONNECT
+echo "GRANT SELECT ON default.t20_0016_table1 TO testuser1" | $BENDSQL_CLIENT_CONNECT
+
+echo '-- ensure the statements not break with PUBLIC role'
+echo "SHOW TABLES;" | $TEST_USER_CONNECT > /dev/null
+echo "SHOW DATABASES;" | $TEST_USER_CONNECT > /dev/null
```
Comment (TCeason) [inline tests/suites/0_stateless/18_rbac/20_0016_rewrite_statements.sh:19]: need tests/suites/0_stateless/18_rbac/20_0016_rewrite_statements.result?

File: tests/suites/0_stateless/18_rbac/20_0012_privilege_access.result (hunk 2 above)
Comment (flaneur2020) [reply tests/suites/0_stateless/18_rbac/20_0012_privilege_access.result:22]: no, i don't think this `user can query view even though he does not have privilige about any table` was the behavior before, this pr is not actually affected the related logic at all.

**Comment by TCeason** [comment]
https://github.com/datafuselabs/databend/pull/14112/files#diff-1c7ca9f866afb39e388942d88ae7da6c2167c9e534134d9ff26b1e7353f7ccceR50-R54

This is crazy, I think we can optimize it in future.

File: src/query/service/src/interpreters/access/privilege_access.rs [hunk 5]
```diff
@@ -209,57 +226,37 @@ impl PrivilegeAccess {
...
+        // skip checking the privilege on system tables.
+        if ((db_name == "system" && SYSTEM_TABLES_ALLOW_LIST.iter().any(|x| x == &table_name))
+            || db_name == "information_schema")
+            && privileges == [UserPrivilegeType::Select]
+        {
+            return Ok(());
         }
+
+        // TODO: remove the verify_ownership parameter
         if verify_ownership {
```
Comment (TCeason) [inline src/query/service/src/interpreters/access/privilege_access.rs:260]: Maybe can modify in this pr?

if verify_ownership  maybe can modify to this:

```suggestion
        if !(db_name=="" && table_name=="") {
```

Comment (TCeason) [reply src/query/service/src/interpreters/access/privilege_access.rs:260]: Better to add next pr.

File: tests/suites/0_stateless/18_rbac/20_0016_rewrite_statements.sh (hunk 4 above)
Comment (TCeason) [reply tests/suites/0_stateless/18_rbac/20_0016_rewrite_statements.sh:19]: Fixed.

File: src/query/service/src/sessions/session_privilege_mgr.rs [hunk 6]
```diff
@@ -265,25 +266,20 @@ impl SessionPrivilegeManager for SessionPrivilegeManagerImpl {
...
-            Some(owner_role) => owner_role,
-            None => return Ok(()),
+        // if the object is not owned by any role, then considered as ACCOUNT_ADMIN, the normal users
+        // can not access it unless ACCOUNT_ADMIN grant relevant privileges to them.
+        let owner_role_name = match role_mgr.find_object_owner(&tenant, object).await? {
+            Some(owner_role) => owner_role.name,
+            None => BUILTIN_ROLE_ACCOUNT_ADMIN.to_string(),
         };
 
         let available_roles = self.get_all_available_roles().await?;
```
Comment (TCeason) [inline src/query/service/src/sessions/session_privilege_mgr.rs:280]: If get err, will return Err(Errcode)

File: src/query/service/src/sessions/session_privilege_mgr.rs [hunk 7]
```diff
@@ -265,25 +266,20 @@ impl SessionPrivilegeManager for SessionPrivilegeManagerImpl {
...
         let role_mgr = RoleCacheManager::instance();
         let tenant = self.session_ctx.get_current_tenant();
 
-        // if the object is not owned by any role, then considered as PUBLIC, which is always true
-        let owner_role = match role_mgr.find_object_owner(&tenant, object).await? {
-            Some(owner_role) => owner_role,
-            None => return Ok(()),
+        // if the object is not owned by any role, then considered as ACCOUNT_ADMIN, the normal users
+        // can not access it unless ACCOUNT_ADMIN grant relevant privileges to them.
+        let owner_role_name = match role_mgr.find_object_owner(&tenant, object).await? {
```
Comment (TCeason) [inline src/query/service/src/sessions/session_privilege_mgr.rs:275]: If find err, will return Err(Errcode)

File: src/query/service/src/interpreters/access/privilege_access.rs [hunk 8]
```diff
@@ -271,13 +268,14 @@ impl PrivilegeAccess {
                         _ => Err(e.add_message("error on validating access")),
                     })?;
             if let Some(object_by_id) = &object_by_id {
-                let result = session.validate_ownership(object_by_id).await;
-                if result.is_ok() {
+                let ok = session.has_ownership(object_by_id).await?;
```
Comment (TCeason) [inline src/query/service/src/interpreters/access/privilege_access.rs:271]: If return Err(Errcode) will return directly.

Comment (TCeason) [reply src/query/service/src/interpreters/access/privilege_access.rs:271]: If has_ownership return Err, validate_access need return Err. This is expected

**Review by TCeason — APPROVED** [review]

**Pull request merged**